eclair-core/src/main/scala/fr/acinq/eclair/wire/FailureMessage.scala

```diff
@@ -93,4 +97,26 @@ object FailureMessageCodecs {
     .typecase(18, ("expiry" | uint32).as[FinalIncorrectCltvExpiry])
     .typecase(19, ("amountMsat" | uint64overflow).as[FinalIncorrectHtlcAmount])
     .typecase(21, provide(ExpiryTooFar))
+
+  /**
+    * Return the failure code for a given failure message. This method actually encodes the failure message, which is a
+    * bit clunky and not particularly efficient. It shouldn't be used on the application's hot path.
+    */
+  def failureCode(failure: FailureMessage): Int = failureMessageCodec.encode(failure).flatMap(uint16.decode).require.value
+
+  /**
+    * An onion-encrypted failure from an intermediate node:
+    * +----------------+----------------------------------+-----------------+----------------------+-----+
+    * | HMAC(32 bytes) | failure message length (2 bytes) | failure message | pad length (2 bytes) | pad |
+    * +----------------+----------------------------------+-----------------+----------------------+-----+
+    * with failure message length + pad length = 256
+    */
+  def failureOnionCodec(mac: Mac32): Codec[FailureMessage] = CommonCodecs.prependmac(
+    paddedFixedSizeBytesDependent(
+      260,
+      "failureMessage" | variableSizeBytes(uint16, FailureMessageCodecs.failureMessageCodec),
+      nBits => {
+        val nBytes = (nBits / 8).toInt
+        variableSizeBytes(uint16, bytes(nBytes - 2)).unit(ByteVector.empty)
```

Comment:
Your implementation works, but I don't know how. You have defined the *zero* as an empty byte vector, but it should be 0x00?

```suggestion
        variableSizeBytes(uint16, bytes(nBytes - 2)).unit(ByteVector.fromByte(0))
```

Reply:
And also it should be:

```suggestion
        variableSizeBytes(uint16, bytes(nBytes - 2).unit(ByteVector.empty))
```

Right?

Reply:
You're right, it worked the same with an empty ByteVector simply because it treats `0x00` as the default value if not specified, but it's better to be explicit.
And it's better to put the unit inside the variableSizeBytes as well, I'll update that.

Reply:
How about: 😎 

```scala
def failureOnionCodec(mac: Mac32): Codec[FailureMessage] = CommonCodecs.prependmac(
    paddedFixedSizeBytesDependent(
      260,
      "failureMessage" | variableSizeBytes(uint16, FailureMessageCodecs.failureMessageCodec),
      nBits => variableSizeBytes(uint16, ignore(nBits - 2 * 8)) // two bytes are used to encode the length of the padding
    ).as[FailureMessage], mac)
```

Reply:
Nitpick:

```scala
def failureOnionCodec(mac: Mac32): Codec[FailureMessage] = CommonCodecs.prependmac(
    paddedFixedSizeBytesDependent(
      260,
      "failureMessage" | variableSizeBytes(uint16, FailureMessageCodecs.failureMessageCodec),
      nBits => "padding" | variableSizeBytes(uint16, ignore(nBits - 2 * 8)) // two bytes are used to encode the length of the padding
    ).as[FailureMessage], mac)
```